src/annotator/plugin/bucket-bar-js.js

```diff
@@ -0,0 +1,52 @@
+import { getBoundingClientRect } from '../highlighter';
+
+// FIXME: Temporary duplication of size constants between here and BucketBar
+const BUCKET_SIZE = 16; // Regular bucket size
+const BUCKET_NAV_SIZE = BUCKET_SIZE + 6; // Bucket plus arrow (up/down)
+const BUCKET_TOP_THRESHOLD = 115 + BUCKET_NAV_SIZE; // Toolbar
+
+/**
+ * Find the closest valid anchor in `anchors` that is offscreen in the direction
+ * indicated.
+ *
+ * @param {Object[]} anchors
+ * @param {'up'|'down'} direction
+ * @return {Object|null} - The closest anchor or `null` if no valid anchor found
+ */
+export function findClosestOffscreenAnchor(anchors, direction) {
+  const { next: closest } = anchors.reduce(
+    (acc, anchor) => {
+      if (!anchor.highlights?.length) {
+        return acc;
+      }
+
+      const top = getBoundingClientRect(anchor.highlights).top;
+
+      if (direction === 'up' && top >= BUCKET_TOP_THRESHOLD) {
+        // We're headed up but the anchor is already below the
+        // visible top of the bucket bar: it's not our guy
+        return acc;
+      } else if (
+        direction === 'down' &&
+        top <= window.innerHeight - BUCKET_NAV_SIZE
+      ) {
+        // We're headed down but this anchor is already above
+        // the usable bottom of the screen: it's not our guy
+        return acc;
+      }
+
+      const dir = direction === 'up' ? +1 : -1;
+      if (acc.next === null || acc.top * dir < top * dir) {
+        // This one is closer than the previous candidate
+        return {
+          next: anchor,
+          top,
+        };
+      } else {
+        return acc;
+      }
+    },
+    { next: null, top: null }
+  );
+  return closest;
+}
```

Comment:
This is a perfectly good translation of the existing code, though I think it could be refactored in future to be more straightforward to read by using a loop rather than an array reduce. A quick sketch:

```js
export function findClosestOffscreenAnchor(anchors, direction) {
  let closestAnchor = null;
  let closestTop = 0;

  for (let anchor of anchors) {
    if (!anchor.highlights?.length) {
      continue;
    }
    const top = getBoundingClientRect(anchor.highlights).top;

    if (direction === 'up' && top >= BUCKET_TOP_THRESHOLD) {
      // Anchor is not off-screen in the "up" direction.
      continue;
    }
    if (direction === 'down' && top <= window.innerHeight - BUCKET_NAV_SIZE) {
      // Anchor is not off-screen in the "down" direction.
      continue;
    }

    if (
      !closestAnchor ||
      (direction === 'up' && top > closestTop) ||
      (direction === 'down' && top < closestTop)
    ) {
      // This is either the first off-screen anchor we've found in the direction
      // we are searching or it is closer to the screen that the previous one.
      closestAnchor = anchor;
      closestTop = top;
    }
  }

  return closestAnchor;
}
```